aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -88,23 +96,69 @@ public void seek(long newPos) {
 
   @Override
   public int read() throws IOException {
-    Preconditions.checkState(!closed, "Cannot read: already closed");
-    positionStream();
+    int[] byteRef = new int[1];
+    try {
+      Tasks.foreach(0)
```

Comment:
This is a lot of duplicate code for the retry logic.  Could we consolidate to a method that wraps just the logic we want to execute?

For example:

```java
retry(() -> {
    Preconditions.checkState(!closed, "Cannot read: already closed");
    positionStream();

    byteRef[0] =  stream.read();
})
```

(you many have to use consumer/function for the range versions, but seems like there might be an more concise way to write this)